### DISCUSSION TITLE
Video length/ frame discrepancies

### DISCUSSION THREAD

**Discussion by meganalo** [bonsai-rx/bonsai#1827] [Q&A]
Hi I’m curious if anyone has had a similar issue to this. 
We have 2 USB 3.0 cameras (Imaging Source DMK 37BUR0234 + their drivers), computer has i9 core, 2 GHz CPU, 16 GB RAM + external RAM.  I am trying to save separate video and csv files for each video source but am running into issues with our current configuration. For one thing, the frame rates reported from each camera are not the same, this is ok if it is a few frames at the beginning or end but the issue is that there is sometimes drift throughout (problem appears to be intermittent), I’m just curious if anyone has found a way to rectify this without a hardware solution. Also, despite having set the timing of the session to 900 seconds the videos only record for a fraction of that time (and the time is not consistent). I've seen a few similar discussions I'm just not sure if anyone has found a solution or if there is an error in my workflow I'm not aware of.

[folder with timestamps and bonsai configuration.zip](https://github.com/bonsai-rx/bonsai/files/15474329/folder.with.timestamps.and.bonsai.configuration.zip)

![image](https://github.com/bonsai-rx/bonsai/assets/128488461/5cf2275c-4544-4d31-afe0-b3866d0f8356)
![image](https://github.com/bonsai-rx/bonsai/assets/128488461/594eba18-01b7-4111-a8ba-2f550a00283c)

**Comment by bruno-f-cruz** [comment]
> I’m just curious if anyone has found a way to rectify this without a hardware solution. 

This will likely be very hard to achieve with USB cameras. The best way to guarantee that two cameras are synchronized is to trigger a frame acquisition via a common hardware trigger.

>  Also, despite having set the timing of the session to 900 seconds the videos only record for a fraction of that time

how did you very this?

**Reply by meganalo** [reply]
Ok, noted about the hardware trigger, thank you. I think we might be able to get away with it in post-processing if we can get the cameras to report back frames somehow.

Good point, my bad I should have specified: the timestamp files include timestamps up to ~15minutes but when the videos get saved they are are much shorter for example 2 min 47 seconds (timing seems random, I had another round that saved as 2 min 24 seconds). Videos from both cameras end up being around the same length though.

**Reply by bruno-f-cruz** [reply]
It seems the encoding frame rate does not match the camera framerate. You can check this by taking the average difference between each timestamp and comparing it to the set encoding frame rate. At the end of the day, trusting the duration of the recorded video usually not a good idea. That duration is just metadata. It usually gets calculated by multiplying encoded fps * number of frames.
